### PR TITLE
Drop `RecentEraAlonzo`

### PR DESCRIPTION
**Pull request by Anviking**
- [x] Drop `RecentEraAlonzo` - we only to support the two most recent eras: `Conway` and `Babbage`.

### Issue Number

ADP-2913

### CODE REVIEW

File: lib/wallet/src/Cardano/Wallet/Write/Tx.hs [hunk 1]
```diff
@@ -746,30 +710,30 @@ utxoFromTxOuts era = withConstraints era $
...
     -> [(TxIn, TxOutInRecentEra)]
-    -> Either ErrInvalidTxOutInEra (Shelley.UTxO (Cardano.ShelleyLedgerEra era))
-utxoFromTxOutsInLatestEra era entries = withConstraints era $ do
-    entries' <- forM entries $ \(i,o) -> do
-        o' <- unwrapTxOutInRecentEra era o
-        return (i, o')
-    return $ Shelley.UTxO $ Map.fromList entries'
-
+    -> (Shelley.UTxO (Cardano.ShelleyLedgerEra era))
+utxoFromTxOutsInLatestEra era entries = withConstraints era $
```
Comment (Anviking) [inline lib/wallet/src/Cardano/Wallet/Write/Tx.hs]: `utxoFromTxOutsInLatestEra` and `utxoFromTxOutsInRecentEra` look very alike 😄

File: lib/wallet/test/unit/Cardano/Wallet/Write/TxSpec.hs [hunk 2]
```diff
@@ -67,10 +67,14 @@ import Test.QuickCheck
 import qualified Cardano.Api as Cardano
 import qualified Cardano.Api.Gen as Cardano
 import qualified Cardano.Ledger.Alonzo.Data as Alonzo
+import Cardano.Ledger.Babbage.TxBody
+    ( BabbageTxOut (..) )
```
Comment (Anviking) [inline lib/wallet/test/unit/Cardano/Wallet/Write/TxSpec.hs]: Oops

File: lib/wallet/test/unit/Cardano/Wallet/Write/TxSpec.hs [hunk 3]
```diff
@@ -67,10 +67,14 @@ import Test.QuickCheck
...
 import qualified Data.Aeson as Aeson
 import qualified Data.Aeson.Types as Aeson
 import qualified Data.ByteString as BS
 import qualified Data.Map as Map
+import Ouroboros.Consensus.Shelley.Eras
+    ( StandardConway )
```
Comment (Anviking) [inline lib/wallet/test/unit/Cardano/Wallet/Write/TxSpec.hs]: Oops

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 4]
```diff
@@ -3094,18 +3094,16 @@ balanceTransaction
...
 
                 mkLedgerUTxO
                     :: [ApiExternalInput n]
                     -> Handler (WriteTx.UTxO (ShelleyLedgerEra era))
                 mkLedgerUTxO ins = do
                     recentEra <- mkRecentEra
-                    liftHandler
-                        . ExceptT
-                        . pure
+                    return
```
Comment (Unisay) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]: I usually use `pure` in this cases at its less powerful but sufficient, as per https://www.lihaoyi.com/post/StrategicScalaStylePrincipleofLeastPower.html#philosophy-principle-of-least-power

File: lib/wallet/src/Cardano/Wallet/Write/Tx.hs [hunk 5]
```diff
@@ -746,30 +710,30 @@ utxoFromTxOuts era = withConstraints era $
...
-            (TxIn, TxOut (ShelleyLedgerEra era))
-    downcast (i, o) = do
-        o' <- unwrapTxOutInRecentEra era o
-        pure (i, o')
+        -> (TxIn, TxOut (ShelleyLedgerEra era))
+    downcast (i, o) =
+        let
+            o' = unwrapTxOutInRecentEra era o
+        in
+            (i, o')
```
Comment (Unisay) [inline lib/wallet/src/Cardano/Wallet/Write/Tx.hs]:
```
    downcast
        :: (TxIn, TxOutInRecentEra)
        -> Either ErrInvalidTxOutInEra (TxIn, TxOut (ShelleyLedgerEra era))
    downcast = traverse (unwrapTxOutInRecentEra era) 
 ```

File: lib/wallet/src/Cardano/Wallet/Write/Tx.hs [hunk 6]
```diff
@@ -746,30 +710,30 @@ utxoFromTxOuts era = withConstraints era $
...
-        return (i, o')
-    return $ Shelley.UTxO $ Map.fromList entries'
-
+    -> (Shelley.UTxO (Cardano.ShelleyLedgerEra era))
+utxoFromTxOutsInLatestEra era entries = withConstraints era $
+    Shelley.UTxO $ Map.fromList $ flip map entries $ \(i,o) ->
+        let
+            o' = unwrapTxOutInRecentEra era o
+        in
+            (i, o')
```
Comment (Unisay) [inline lib/wallet/src/Cardano/Wallet/Write/Tx.hs]:
```suggestion
    Shelley.UTxO $ Map.fromList $ fmap (unwrapTxOutInRecentEra era) <$> entries
```

Comment (Anviking) [reply lib/wallet/src/Cardano/Wallet/Write/Tx.hs]: Is that the `Functor (,)` instance? I'd prefer to rely on `Control.Arrow.second` in that case.

File: lib/wallet/src/Cardano/Wallet/Write/Tx.hs (hunk 1 above)
Comment (Anviking) [reply lib/wallet/src/Cardano/Wallet/Write/Tx.hs]: Dropped `utxoFromTxOutsInLatestEra `

File: specifications/api/swagger.yaml [hunk 7]
```diff
@@ -6174,8 +6152,6 @@ x-responsesBalanceTransaction: &responsesBalanceTransaction
             - <<: *errTranslationError
             - <<: *errUnresolvedInputs
             - <<: *errInputResolutionConflicts
-            - <<: *errBalanceTxInlineScriptsNotSupportedInAlonzo
-            - <<: *errBalanceTxInlineDatumsNotSupportedInAlonzo
```
Comment (Anviking) [inline specifications/api/swagger.yaml:6178]: Dropping the errors here in `swagger.yaml` as well.

File: lib/wallet/src/Cardano/Wallet/Write/Tx.hs (hunk 5 above)
Comment (Anviking) [reply lib/wallet/src/Cardano/Wallet/Write/Tx.hs]: Went with `second` in the other function instead, and the `Either` is gone now.

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs (hunk 4 above)
Comment (Anviking) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]: done

**Review by Unisay — APPROVED** [review]
Nice!

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs [hunk 8]
```diff
@@ -477,15 +476,6 @@ instance IsServerError ErrDecodeTx where
...
 
-instance IsServerError WriteTx.ErrInvalidTxOutInEra where
-     toServerError = \case
-         WriteTx.ErrInlineDatumNotSupportedInAlonzo ->
-             apiError err400 BalanceTxInlineDatumsNotSupportedInAlonzo
-                 "Inline datums are not supported in the Alonzo era."
-         WriteTx.ErrInlineScriptNotSupportedInAlonzo ->
-             apiError err400 BalanceTxInlineScriptsNotSupportedInAlonzo
-                 "Inline scripts are not supported in the Alonzo era."
-
```
Comment (Anviking) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs:488]: If a user were able to hit these errors today, before this PR, they'd now hit `Write.ErrOldEraNotSupported` instead.

**Comment by Anviking** [comment]
bors r-

```
cardano-wallet-test-unit> Failures:
--
  | cardano-wallet-test-unit>   test/unit/Cardano/Wallet/Api/TypesSpec.hs:973:9:
  | cardano-wallet-test-unit>   1) Cardano.Wallet.Api.Types, Api Errors, Every ApiErrorInfo constructor has a corresponding schema type
  | cardano-wallet-test-unit>        Falsified (after 1 test):
  | cardano-wallet-test-unit>          Missing ApiErrorInfo constructors for:
  | cardano-wallet-test-unit>          ["BalanceTxInlineDatumsNotSupportedInAlonzo","BalanceTxInlineScriptsNotSupportedInAlonzo"]
  | cardano-wallet-test-unit>          Each of these need a corresponding swagger type of the form:
  | cardano-wallet-test-unit>          x-errConstructorName
  | cardano-wallet-test-unit>   To rerun use: --match "/Cardano.Wallet.Api.Types/Api Errors/Every ApiErrorInfo constructor has a corresponding schema type/"
  | cardano-wallet-test-unit>   test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:2467:5:
  | cardano-wallet-test-unit>   2) Cardano.Wallet.Shelley.Transaction.balanceTransaction produces valid transactions or fails
  | cardano-wallet-test-unit>        Falsified (after 131 tests and 75 shrinks):
  | cardano-wallet-test-unit>          Wallet:
  | cardano-wallet-test-unit>            assumptions: - scriptTemplate:
  | cardano-wallet-test-unit>            changeAddressGen:
  | cardano-wallet-test-unit>              - changeAddr0: Addr Mainnet (KeyHashObj (KeyHash "e718155067792092dadf79551e28407f8061b8201ce0a047bba26c5c")) (StakeRefBase (KeyHashObj (KeyHash "ca5aea94925bc3e2148f3eca1a311286705c8be8fcfd284e8f3c7eb0")))
  | cardano-wallet-test-unit>              - max address length: 57
  | cardano-wallet-test-unit>            utxo: []
  | cardano-wallet-test-unit>          PartialTx:
  | cardano-wallet-test-unit>            inputs: - (TxIn "01f4b788593d4f70de2a45c2e1e87088bfbdfa29577ae1b62aba60e095e3ab53" (TxIx 11),TxOut (AddressInEra (ShelleyAddressInEra ShelleyBasedEraBabbage) (ShelleyAddress Mainnet (ScriptHashObj (ScriptHash "c7325ad987435b0fe9090979b483f67bafe9cd8d07cec9c29792d6a2")) (StakeRefBase (KeyHashObj (KeyHash "61e5ca65aec4b27dfcf670cf7609162f45309645f267bbfdd4f73986"))))) (TxOutValue MultiAssetInBabbageEra (valueFromList [])) (TxOutDatumHash ScriptDataInBabbageEra "03170a2e7597b7b7e3d84c05391d139a62b157e78786d8c082f29dcf4c111314") ReferenceScriptNone)
  | cardano-wallet-test-unit>            redeemers: []
  | cardano-wallet-test-unit>            tx:
  | cardano-wallet-test-unit>              ShelleyTx ShelleyBasedEraBabbage
  | cardano-wallet-test-unit>                  ( AlonzoTx
  | cardano-wallet-test-unit>                      { body = TxBodyConstr TxBodyRaw
  | cardano-wallet-test-unit>                          { _spendInputs = fromList
  | cardano-wallet-test-unit>                              [ TxIn
  | cardano-wallet-test-unit>                                  ( TxId
  | cardano-wallet-test-unit>                                      { _unTxId = SafeHash "01f4b788593d4f70de2a45c2e1e87088bfbdfa29577ae1b62aba60e095e3ab53" }
  | cardano-wallet-test-unit>                                  )
  | cardano-wallet-test-unit>                                  ( TxIx 11 )
  | cardano-wallet-test-unit>                              ]
  | cardano-wallet-test-unit>                          , _collateralInputs = fromList []
  | cardano-wallet-test-unit>                          , _referenceInputs = fromList []
  | cardano-wallet-test-unit>                          , _outputs = StrictSeq
  | cardano-wallet-test-unit>                              { fromStrict = fromList
  | cardano-wallet-test-unit>                                  [ Sized
  | cardano-wallet-test-unit>                                      { sizedValue =
  | cardano-wallet-test-unit>                                          ( AddrBootstrap
  | cardano-wallet-test-unit>                                              ( BootstrapAddress
  | cardano-wallet-test-unit>                                                  ( Address
  | cardano-wallet-test-unit>                                                      { addrRoot = c2bf8fb30dc17a1ecb3b02132a5109136c8233e96f06a85684dd9f03
  | cardano-wallet-test-unit>                                                      , addrAttributes = Attributes
  | cardano-wallet-test-unit>                                                          { data_ = AddrAttributes
  | cardano-wallet-test-unit>                                                              { aaVKDerivationPath = Nothing
  | cardano-wallet-test-unit>                                                              , aaNetworkMagic = NetworkMainOrStage
  | cardano-wallet-test-unit>                                                              }
  | cardano-wallet-test-unit>                                                          }
  | cardano-wallet-test-unit>                                                      , addrType = ATVerKey
  | cardano-wallet-test-unit>                                                      }
  | cardano-wallet-test-unit>                                                  )
  | cardano-wallet-test-unit>                                              )
  | cardano-wallet-test-unit>                                          , MaryValue 4294708831
  | cardano-wallet-test-unit>                                              ( fromList [] )
  | cardano-wallet-test-unit>                                          , DatumHash
  | cardano-wallet-test-unit>                                              ( SafeHash "03170a2e7597b7b7e3d84c05391d139a62b157e78786d8c082f29dcf4c111314" )
  | cardano-wallet-test-unit>                                          , SNothing
  | cardano-wallet-test-unit>                                          )
  | cardano-wallet-test-unit>                                      , sizedSize = 90
  | cardano-wallet-test-unit>                                      }
  | cardano-wallet-test-unit>                                  ]
  | cardano-wallet-test-unit>                              }
  | cardano-wallet-test-unit>                          , _collateralReturn = SNothing
  | cardano-wallet-test-unit>                          , _totalCollateral = SNothing
  | cardano-wallet-test-unit>                          , _certs = StrictSeq
  | cardano-wallet-test-unit>                              { fromStrict = fromList [] }
  | cardano-wallet-test-unit>                          , _wdrls = Wdrl
  | cardano-wallet-test-unit>                              { unWdrl = fromList
  | cardano-wallet-test-unit>                                  [
  | cardano-wallet-test-unit>                                      ( RewardAcnt
  | cardano-wallet-test-unit>                                          { getRwdNetwork = Mainnet
  | cardano-wallet-test-unit>                                          , getRwdCred = ScriptHashObj
  | cardano-wallet-test-unit>                                              ( ScriptHash "a646474b8f5431261506b6c273d307c7569a4eb6c96b42dd4a29520a" )
  | cardano-wallet-test-unit>                                          }
  | cardano-wallet-test-unit>                                      , Coin 0
  | cardano-wallet-test-unit>                                      )
  | cardano-wallet-test-unit>                                  ,
  | cardano-wallet-test-unit>                                      ( RewardAcnt
  | cardano-wallet-test-unit>                                          { getRwdNetwork = Mainnet
  | cardano-wallet-test-unit>                                          , getRwdCred = KeyHashObj
  | cardano-wallet-test-unit>                                              ( KeyHash "cef0b0176bb5621a122a4ecf84776ef8d361b2cd1c97556b92831463" )
  | cardano-wallet-test-unit>                                          }
  | cardano-wallet-test-unit>                                      , Coin 4294891360
  | cardano-wallet-test-unit>                                      )
  | cardano-wallet-test-unit>                                  ]
  | cardano-wallet-test-unit>                              }
  | cardano-wallet-test-unit>                          , _txfee = Coin 0
  | cardano-wallet-test-unit>                          , _vldt = ValidityInterval
  | cardano-wallet-test-unit>                              { invalidBefore = SJust
  | cardano-wallet-test-unit>                                  ( SlotNo 18446744073709547246 )
  | cardano-wallet-test-unit>                              , invalidHereafter = SJust
  | cardano-wallet-test-unit>                                  ( SlotNo 18446744073709547777 )
  | cardano-wallet-test-unit>                              }
  | cardano-wallet-test-unit>                          , _update = SNothing
  | cardano-wallet-test-unit>                          , _reqSignerHashes = fromList []
  | cardano-wallet-test-unit>                          , _mint = MaryValue 0
  | cardano-wallet-test-unit>                              ( fromList [] )
  | cardano-wallet-test-unit>                          , _scriptIntegrityHash = SJust
  | cardano-wallet-test-unit>                              ( SafeHash "c5b8c05c64433aff798114007d910fbdfcd3d402d6a07f15fc2e2ba2adda10bc" )
  | cardano-wallet-test-unit>                          , _adHash = SJust
  | cardano-wallet-test-unit>                              ( AuxiliaryDataHash
  | cardano-wallet-test-unit>                                  { unsafeAuxiliaryDataHash = SafeHash "0fb595b675ef3e08564e3fb91c1ce37d4c4ce7d3e22317e34500973e2575c52b" }
  | cardano-wallet-test-unit>                              )
  | cardano-wallet-test-unit>                          , _txnetworkid = SNothing
  | cardano-wallet-test-unit>                          }
  | cardano-wallet-test-unit>                      , wits = TxWitnessRaw
  | cardano-wallet-test-unit>                          { _txwitsVKey = fromList []
  | cardano-wallet-test-unit>                          , _txwitsBoot = fromList []
  | cardano-wallet-test-unit>                          , _txscripts = fromList
  | cardano-wallet-test-unit>                              [
  | cardano-wallet-test-unit>                                  ( ScriptHash "a646474b8f5431261506b6c273d307c7569a4eb6c96b42dd4a29520a"
  | cardano-wallet-test-unit>                                  , PlutusScript PlutusV2 ScriptHash "a646474b8f5431261506b6c273d307c7569a4eb6c96b42dd4a29520a"
  | cardano-wallet-test-unit>                                  )
  | cardano-wallet-test-unit>                              ]
  | cardano-wallet-test-unit>                          , _txdats = TxDatsRaw
  | cardano-wallet-test-unit>                              ( fromList [] )
  | cardano-wallet-test-unit>                          , _txrdmrs = RedeemersRaw
  | cardano-wallet-test-unit>                              ( fromList [] )
  | cardano-wallet-test-unit>                          }
  | cardano-wallet-test-unit>                      , isValid = IsValid True
  | cardano-wallet-test-unit>                      , auxiliaryData = SNothing
  | cardano-wallet-test-unit>                      }
  | cardano-wallet-test-unit>                  )
  | cardano-wallet-test-unit>          StdGenSeed 28
  | cardano-wallet-test-unit>          balanceTransaction failed: ErrBalanceTxAssignRedeemers (ErrAssignRedeemersTranslationError (ByronTxOutInContext (TxOutFromOutput (TxIx 0))))
  | cardano-wallet-test-unit>   To rerun use: --match "/Cardano.Wallet.Shelley.Transaction/balanceTransaction/produces valid transactions or fails/"
  | cardano-wallet-test-unit> Randomized with seed 325890028
  | cardano-wallet-test-unit> Finished in 236.2640 seconds, used 941.0023 seconds of CPU time
  | cardano-wallet-test-unit> 4750 examples, 2 failures, 29 pending


```

**Comment by Anviking** [comment]
bors try

**Pull request merged**